pkg/porter/version.go

```
@@ -1,80 +1,111 @@
 	}
-}
 
-func (system *SystemInfo) printSystemInfo(out io.Writer) error {
-	_, err := fmt.Fprintf(out, "os: %s\n", system.OS)
-	_, err = fmt.Fprintf(out, "arch: %s\n", system.Arch)
-	return err
+	return version.PrintVersion(p.Context, opts.Options, metadata)
 }
 
-func printSectionHeader(out io.Writer, header string){
-	underline := "-------"
-	_, _ = fmt.Fprintf(out, "\n%s\n%s\n", header, underline)
-
+func getSystemInfo() *SystemInfo {
+	return &SystemInfo{
+		OS:   runtime.GOOS,
+		Arch: runtime.GOARCH,
+	}
 }
 
-func (p *Porter) PrintDebugInfo(opts version.Options) error {
-	// force opts to print version as plaintext
+func (p *Porter) PrintDebugInfo(ctx *context.Context, opts VersionOpts, versionMetadata mixin.Metadata) error {
 	opts.RawFormat = string(printer.FormatPlaintext)
-	err := p.PrintVersion(opts)
+	mixins, err := p.ListMixins()
+	sysInfo := getSystemInfo()
 	if err != nil {
-		return errors.Wrap(err, "Failed to print version")
+		_ = errors.Wrap(err, "Failed to get list of mixins")
 	}
-
-	printSectionHeader(p.Context.Out, "System")
-	systemInfo := getSystemInfo()
-	err = systemInfo.printSystemInfo(p.Out)
-	if err != nil {
-		return errors.Wrap(err, "Failed to print system information")
+	sysDebugInfo := SystemDebugInfo{
+		Version: versionMetadata,
+		SysInfo: *sysInfo,
+		Mixins:  mixins,
 	}
-	printSectionHeader(p.Context.Out, "Mixins")
 
-	mixins, err := p.ListMixins()
-	if err != nil {
+	switch opts.Format {
+	case printer.FormatJson:
+		return printer.PrintJson(ctx.Out, sysDebugInfo)
+	case printer.FormatPlaintext:
+		plaintextTmpl := `{{.Version.Name}} {{.Version.VersionInfo.Version}} ({{.Version.VersionInfo.Commit}})
+
+System
+-------
+os: {{.SysInfo.OS}}
+arch: {{.SysInfo.Arch}}
+{{if .Mixins}}
+Mixins
+-------
+{{.Mixins.PrintMixinsTable}}{{end}}
+`
+		tmpl, err := template.New("systemDebugInfo").Parse(plaintextTmpl)
+		if err != nil {
+			_ = errors.Wrap(err, "Failed to print system debug information")
```

At this point, if we can't parse the template, then we should return the error, and not ignore it with `_` because it's an unrecoverable error. 

If the template isn't parsed, then continuing on to call Execute on the template (which is nil when err is populated) will cause a nil reference exception.